### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mecha**

### CODE REVIEW

File: src/EddBkLicensingModule.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        if ($c === null) {
+            return;
+        }
+
+        if (\is_admin()) {
+            /*
+             * Retrieving the license from the container instantiates it.
+             * Unfortunately, upon construction it registers all the necessary hooks
+             * So this is enough to initialize the "licensing" for EDDBK.
+             */
```
Comment (XedinUnknown) [inline src/EddBkLicensingModule.php:107]: Tough luck... :(

**Review by XedinUnknown — APPROVED** [review]
This looks good. However, I have a question:

Is this really enough to have the licensing work? Do we need some of the abstractions we had made in the past for WPRA?

**Comment by mecha** [comment]
Yes. EDD handles most of the work regarding UI, as well as sending and handling requests. All we have to do is register the `EDD_License`. In many ways except for architecture, it is equivalent to our WPRA abstraction, where our addons are only required to register themselves with Core.

**Comment by XedinUnknown** [comment]
From what I can see, this is nothing like the WPRA abstraction. In fact, while examining that abstraction, I got more questions:

1. Where does one enter the license key?
2. Where is the license key stored/retrieved from?
3. Where does the EDD licensing lib come from?

**Comment by mecha** [comment]
1. In the EDD Settings page, there's a Licensing tab just like WPRA
2. EDD stores it in the WP settings table
3. EDD

**Comment by XedinUnknown** [comment]
So, WRT point 3, check out the [EDD Software Licensing unofficial repo][edd-software-licensing].

[edd-software-licensing]: https://github.com/wp-premium/edd-software-licensing

**Comment by mecha** [comment]
I don't think you understand. Software Licensing is not required for our plugin to have licensing functionality. EDD, the dependency plugin, is what provides the `EDD_License` class. Software Licensing is the server part of the licensing system, which RebelCode will host on their licensing server.

In truth, we could do all the licensing work ourselves, but EDD provides both the UI and the HTTP interactions for its add-ons, both official and unofficial (which EDDBK is).

**Comment by XedinUnknown** [comment]
Yea, no problem. But nothing in this repo defines the used `EDD_License` class, so I was looking for a package that we could at least dev-depend on. Apparently, I got the wrong package.

**Comment by XedinUnknown** [comment]
WRT point 1, how does EDD licensing settings relate to EDDBK licensing settings?

**Comment by XedinUnknown** [comment]
Maybe it's [this one](https://github.com/easydigitaldownloads/EDD-License-handler)?

**Comment by mecha** [comment]
Yeah I thought so too. But it's out-dated as hell. EDD's [licensing handler](https://github.com/easydigitaldownloads/easy-digital-downloads/blob/master/includes/class-edd-license-handler.php) was modified 4 months ago. The one you linked was last modified in 2015.


> WRT point 1, how does EDD licensing settings relate to EDDBK licensing settings?

I should probably explain the entire licensing process, since it's evident that you're a bit confused. Hopefully this will avoid any further back and forth.

WordPress by default checks for, and provides, updates from `wordpress.org`. This is filterable, and EDD's license handler filters this to use a different server. That's why the [plugin main file][1] and [server URL][2] are required. The plugin file is used in the WP filter to detect the plugin being updated, and the server URL is used as the filter result.

Software Licensing can be used on the server to listen to requests made for checking latest version and sending zip files for updates. However, it has one restriction that `wordpress.org` does not; it only provides updates if a valid license is included in the request.

For this, the client needs to send that license. That's where EDD's licensing handler comes in. It provides UI in its settings page for entering and activating licenses, and sends the license key to the server when a version check or update is needed.

We use the same license handler in WPRA. But in WPRA, the Core doesn't need licensing and the UI needs to be located in the WPRA settings page. So we added additional abstraction for WPRA addons to register themselves and license handlers, as well as the UI elements, are created for them automatically by Core.

[1]: https://github.com/RebelCode/rcmod-eddbk-licensing/blob/d8112044814da8f24b82d05178c9a31933016f20/services.php#L8

[2]: https://github.com/RebelCode/rcmod-eddbk-licensing/blob/d8112044814da8f24b82d05178c9a31933016f20/services.php#L13

**Comment by XedinUnknown** [comment]
So, WRT point 1, just confirm whether or not this is true: EDD provides a UI for EDDBK's licensing.

**Comment by mecha** [comment]
Yes, it does. It can be found in the EDD settings page. If at least one EDD addon is active, a _Licensing_ tab will appear.

> It provides UI in its settings page for entering and activating licenses

**Comment by XedinUnknown** [comment]
And EDD knows about the add-ons because there's a `EDD_License` instance registered for it?

**Comment by XedinUnknown** [comment]
Also, see easydigitaldownloads/EDD-License-handler#43.

**Comment by mecha** [comment]
> And EDD knows about the add-ons because there's a EDD_License instance registered for it?

Yep. More specifically, because an `EDD_License` is instantiated. It's constructor will set up all the right hooks.

**Comment by XedinUnknown** [comment]
So, hopefully the final question on this PR: would it be possible to dev-depend on the outdated package? Like this we can at least give a hint of what is required. Maybe the constructor signature, if it's the same. And easily updatable in the future, should EDD decide to bring their package up to date.

**Comment by mecha** [comment]
The new version accepts a new argument (last in the signature) which we are not using. So I'd say dev-requiring it is definitely possible. It shouldn't conflict in production since it won't be installed at all.

**Comment by XedinUnknown** [comment]
Exactly. I was hoping for this. <3 Go. Do.

**Comment by mecha** [comment]
Apparently, not even that package is on packagist 😢 

https://packagist.org/?query=edd%20license

**Comment by XedinUnknown** [comment]
@mecha, how about [this](https://getcomposer.org/doc/05-repositories.md#package-2)?

**Comment by mecha** [comment]
That would work. But is it really ok to include a dependent's package information in this package? Seems like a hack. I'd rather wait for EDD to potentially add the package themselves, since there's no rush for this particular module.

**Comment by XedinUnknown** [comment]
I would prefer if there was some link established between code that consumes `EDD_License`, and code that defines it, before the licensing module is released.

**Comment by mecha** [comment]
@XedinUnknown Added the package!

File: composer.json [hunk 2]
```diff
@@ -30,5 +31,31 @@
     "scripts": {
         "test": "phpunit",
         "csfix": "php-cs-fixer fix -vvv"
-    }
+    },
+    "repositories": [
+        {
+            "type": "package",
+            "package": {
+                "name": "easydigitaldownloads/edd-License-handler",
```
Comment (XedinUnknown) [inline composer.json:39]: Wow such nice.

**Review by XedinUnknown — APPROVED** [review]
So, EDD has neither updated their SL repo, nor replied in the issue thread. Therefore, for now we carry on by manually adding the link to their repo, so that at least the method signatures become apparent.

**Pull request merged**